src/components/AnimateSharedLayout/stack.ts

```diff
@@ -241,3 +248,62 @@ export class LayoutStack {
         return this.lead?.config.transition
     }
 }
+
+export class TreeStack extends Stack<LayoutTree> {
+    sortOrder() {
+        this.order.sort((a, b) => a.props.layoutOrder - b.props.layoutOrder)
+    }
+
+    add(child: LayoutTree) {
+        if (child.layoutOrder === undefined) return
+
+        const { layoutOrder } = child.props
+        let index = (this.order as LayoutTree[]).findIndex(
+            stackChild => layoutOrder <= (stackChild.layoutOrder || 0)
+        )
+
+        if (index === -1) index = this.order.length
+
+        this.order.splice(index, 0, child)
+    }
+
+    /**
+     * Generate LayoutStacks from the children of the Lead and Follow trees.
+     */
+    createLayoutStacks(): Map<string, LayoutStack> {
+        const stacks: Map<string, LayoutStack> = new Map()
+
+        if (this.follow) {
+            for (const child of this.follow.children) {
+                if (!child.layoutId) continue
+
+                const stack = new LayoutStack()
+
+                child.isPresent = this.follow.isPresent
+                stack.add(child)
+                stacks.set(child.layoutId, stack)
+                stack.updateLeadAndFollow()
+            }
+        }
+
+        if (this.lead) {
+            for (const child of this.lead.children) {
+                if (!child.layoutId) continue
+
+                let stack = stacks.get(child.layoutId)
+                const existingStack = !!stack
+
+                stack = stack || new LayoutStack()
+
+                child.isPresent = this.lead.isPresent
+
+                stack.add(child)
+                stack.updateLeadAndFollow()
```

Comment:
Calling updateLeadAndFollow seems to make sense here since it allows us to avoid another loop though